### PR TITLE
Adding a new template to connect VCNs using multiple VNICs

### PR DESCRIPTION
**Pull request by prasanna-naik-oracle**
VCNs with non-overlapping subnets can talk to each other using attached VNICs. This template shows the same.

### CODE REVIEW

File: docs/solutions/connect_vcns_using_multiple_vnics/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+***
+This example creates 2 VCNs with non-overlapping subnets and establishes connectivity between the 2 using attached VNICs.
+
+Each VCN has a public subnet and a private subnet. Each subnet is created with a separate security list and route table. 
+The template then launches a private instance in each one of the private subnets.
+
+A public instance is created in the public subnet of the first VCN. 
+The public instance is configured as a Bridge instance (by enabling and configuring firewall to do forwarding).
+
+The first VCN's private subnet's route table is configured to use the Bridge instance's private IP address as default route target. See [Using a Private IP as a Route Target](https://docs.us-phoenix-1.oraclecloud.com/Content/Network/Tasks/managingroutetables.htm#privateip) for more details on this feature.
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/README.md]: "the default route target"

File: docs/solutions/connect_vcns_using_multiple_vnics/env-vars [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export TF_VAR_tenancy_ocid=""
+export TF_VAR_user_ocid=""
+export TF_VAR_fingerprint=""
+export TF_VAR_private_key_path=""
+
+### Region
+export TF_VAR_region=""
+
+### AD
+export TF_VAR_AD=""
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/env-vars]: Looks like this AD var is 1,2,3 (as opposed to zero-based of the full AD name). Should comment on that.

File: docs/solutions/connect_vcns_using_multiple_vnics/env-vars [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+### Compartment
+export TF_VAR_compartment_ocid=""
+
+### VCN-1 configuration
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/env-vars]: I would remove all of these from vcn_cidr to InstanceOSVersion. The config sets good defaults for these, but if they are in your env-vars like this then they will get set to "" and the apply will fail. I suspect that most users will be fine with these defaults for an example like this.

File: docs/solutions/connect_vcns_using_multiple_vnics/output.tf [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+output "PrivateInstance1 Private IP" {
+    value = "${oci_core_instance.PrivateInstance.private_ip}"
+}
+
+output "PrivateInstance2 Private IP" {
+    value = "${oci_core_instance.PrivateInstance2.private_ip}"
+}
+
+output "SSH login to the Bridge Instance" {
+    value = "ssh -A opc@${data.oci_core_vnic.BridgeInstanceVnic1.public_ip_address}"
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/output.tf]: These instructions in the output are great. Do we want to use -A here though? I was able to login into instance 1 with the following, after setting the variables in the command:
ssh -i $PRIVATE_KEY_PATH -o ProxyCommand="ssh -i $PRIVATE_KEY_PATH opc@$BASTION_IP -W %h:%p %r" opc@$PRIVATE_INSTANCE_IP

File: docs/solutions/connect_vcns_using_multiple_vnics/vcn1.tf [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+variable "mgmt_subnet_cidr" {
+    default = "10.0.0.0/24"
+}
+
+variable "private_subnet_cidr" {
+    default = "10.0.1.0/24"
+}
+
+provider "oci" {
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/vcn1.tf]: nitpick, but usually in larger examples like this the provider is in provider.tf and the variables in variables.tf.

File: docs/solutions/connect_vcns_using_multiple_vnics/vcn2.tf [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    compartment_id = "${var.compartment_ocid}"
+    display_name = "MgmtSecurityList2"
+    vcn_id = "${oci_core_virtual_network.CoreVCN2.id}"
+
+    egress_security_rules = [{
+        protocol = "all"
+        destination = "0.0.0.0/0"
+    }]
+
+    ingress_security_rules = [{
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/vcn2.tf]: Any reason to allow 80/443 in from anywhere?

File: docs/solutions/connect_vcns_using_multiple_vnics/env-vars (hunk 3 above)
Comment (prasanna-naik-oracle) [reply docs/solutions/connect_vcns_using_multiple_vnics/env-vars]: Wanted to give an option for users to change this in the env-vars if they wanted to.

File: docs/solutions/connect_vcns_using_multiple_vnics/output.tf (hunk 4 above)
Comment (prasanna-naik-oracle) [reply docs/solutions/connect_vcns_using_multiple_vnics/output.tf]: From bridge instance, they have to login to the private instance for which forwarding with -A is required.

File: docs/solutions/connect_vcns_using_multiple_vnics/vcn2.tf (hunk 6 above)
Comment (prasanna-naik-oracle) [reply docs/solutions/connect_vcns_using_multiple_vnics/vcn2.tf]: Removed from both vcn1 and vcn2.tf

**Comment by prasanna-naik-oracle** [comment]
Addressed all the review comments.

File: docs/solutions/connect_vcns_using_multiple_vnics/README.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+* Update env-vars with the required information. Most examples use the same set of environment variables so you only need to do this once.
+* Source env-vars
+  * `$ . env-vars`
+* Update variables in bridge.tf as applicable to your target environment.
+
+Once the environment is built, both the private instances in the different VCNs should be connected. You can login to the public Bridge instance, from there login to the private instances and then ping/SSH the other private instance to verify conectivity between the private instances in different VCNs.
+
+### How to validate this example
+Steps to validate this example will also be listed in the output when the terraform is deployed. 
+1. Enable ssh forwarding from your machine by performing ""ssh-add ~/.ssh/id_rsa"
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/README.md]: nitpick: extra quote

File: docs/solutions/connect_vcns_using_multiple_vnics/env-vars (hunk 3 above)
Comment (briangustafson) [reply docs/solutions/connect_vcns_using_multiple_vnics/env-vars]: How about commenting these out and mark optional, like:
```
######
# Optional Variables
######

### AD
# Possible values are 1, 2, 3
# export TF_VAR_AD=""

### VCN-1 configuration
# export TF_VAR_vcn_cidr=""
# export TF_VAR_mgmt_subnet_cidr=""
# export TF_VAR_private_subnet_cidr=""

### VCN-2 configuration
# export TF_VAR_vcn_cidr2=""
# export TF_VAR_mgmt_subnet_cidr2=""
# export TF_VAR_private_subnet_cidr2=""

### Instance configuration
# export TF_VAR_InstanceShape=""
# export TF_VAR_InstanceOS=""
# export TF_VAR_InstanceOSVersion=""
```

File: docs/solutions/connect_vcns_using_multiple_vnics/vcn1.tf [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+    egress_security_rules = [{
+        protocol = "all"
+        destination = "0.0.0.0/0"
+    }]
+
+    ingress_security_rules = [{
+        protocol = "all"
+        source = "${var.vcn_cidr}"
+    },
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/vcn1.tf]: nitpick: formatting is a little strange... can you run "terraform fmt" in this directory?

File: docs/solutions/connect_vcns_using_multiple_vnics/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+2. Login to Bridge instance using "ssh -A <Bridge-Instance-Public-IP>"
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/README.md]: These commands are getting messed up when viewing in github.... try the "view the whole file" button on this file.

File: docs/solutions/connect_vcns_using_multiple_vnics/vcn1.tf [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+resource "oci_core_security_list" "PrivateSecurityList" {
+    compartment_id = "${var.compartment_ocid}"
+    display_name = "PrivateSecurityList"
+    vcn_id = "${oci_core_virtual_network.CoreVCN.id}"
+    egress_security_rules = [{
+        protocol = "all"
+        destination = "0.0.0.0/0"
+    }]
+    ingress_security_rules = [{
+        protocol = "6"
```
Comment (briangustafson) [inline docs/solutions/connect_vcns_using_multiple_vnics/vcn1.tf]: This rule isn't really needed, since it is covered by the second rule, right?

**Comment by prasanna-naik-oracle** [comment]
Incorporated all review comments. Thanks for reviewing. 
Ran "terraform fmt" on the whole folder, so many files will have formatting update changes.

**Pull request merged**